### ISSUE TITLE
[dynamo][GPU] numerics failure for autograd iree_turbine backend

### ISSUE THREAD

**Issue by zjgarvey** [iree-org/iree-turbine#954]
In #953 , I'm adding a generic `torch.compile` backend which should work for any device ("cuda"/"cpu"), and which should enable `CustomOp` expansion. 

When adding some basic tests, (see `tests/dynamo/backend_test.py`) I noticed that the compiled **forward** pass of a function had incorrect numerics, but **only** when input tensors require gradients. Some intermediate results need to be returned by IREE during the forward pass, in order to be saved for the backward pass, and this is the only difference in the IR. 

Note: this numerics quirk is not present for CPU.

### Forward IR with `requires_grad=True`

```mlir
#map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
module {
  util.func public @main$async(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view, %arg3: !hal.buffer_view, %arg4: !hal.fence, %arg5: !hal.fence) -> (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view) attributes {inlining_policy = #util.inline.never, iree.abi.model = "coarse-fences", iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = hal.tensor.import wait(%arg4) => %arg0 : !hal.buffer_view -> tensor<2xf32>
    %1 = hal.tensor.import wait(%arg4) => %arg1 : !hal.buffer_view -> tensor<2x3x1x1xf32>
    %2 = hal.tensor.import wait(%arg4) => %arg2 : !hal.buffer_view -> tensor<2x3x16x16xf32>
    %3 = hal.tensor.import wait(%arg4) => %arg3 : !hal.buffer_view -> tensor<2x2x16x16xf32>
    %4 = tensor.empty() : tensor<2x2x16x16xf32>
    %broadcasted = linalg.broadcast ins(%0 : tensor<2xf32>) outs(%4 : tensor<2x2x16x16xf32>) dimensions = [0, 2, 3] 
    %5 = linalg.conv_2d_nchw_fchw {dilations = dense<1> : vector<2xi64>, strides = dense<1> : vector<2xi64>} ins(%2, %1 : tensor<2x3x16x16xf32>, tensor<2x3x1x1xf32>) outs(%broadcasted : tensor<2x2x16x16xf32>) -> tensor<2x2x16x16xf32>
    %6 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%5 : tensor<2x2x16x16xf32>) outs(%4 : tensor<2x2x16x16xf32>) {
    ^bb0(%in: f32, %out: f32):
      %14 = arith.cmpf ugt, %in, %cst : f32
      %15 = arith.select %14, %in, %cst : f32
      linalg.yield %15 : f32
    } -> tensor<2x2x16x16xf32>
    %7 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%6, %3 : tensor<2x2x16x16xf32>, tensor<2x2x16x16xf32>) outs(%4 : tensor<2x2x16x16xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %14 = arith.mulf %in, %in_0 : f32
      linalg.yield %14 : f32
    } -> tensor<2x2x16x16xf32>
    %8:5 = hal.tensor.barrier join(%7, %1, %2, %3, %6 : tensor<2x2x16x16xf32>, tensor<2x3x1x1xf32>, tensor<2x3x16x16xf32>, tensor<2x2x16x16xf32>, tensor<2x2x16x16xf32>) => %arg5 : !hal.fence
    %9 = hal.tensor.export %8#0 : tensor<2x2x16x16xf32> -> !hal.buffer_view
    %10 = hal.tensor.export %8#1 : tensor<2x3x1x1xf32> -> !hal.buffer_view
    %11 = hal.tensor.export %8#2 : tensor<2x3x16x16xf32> -> !hal.buffer_view
    %12 = hal.tensor.export %8#3 : tensor<2x2x16x16xf32> -> !hal.buffer_view
    %13 = hal.tensor.export %8#4 : tensor<2x2x16x16xf32> -> !hal.buffer_view
    util.return %9, %10, %11, %12, %13 : !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view
  }
  util.func public @main(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view, %arg3: !hal.buffer_view) -> (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view) attributes {iree.abi.stub} {
    %0 = util.null : !hal.fence
    %c-1_i32 = arith.constant -1 : i32
    %c0 = arith.constant 0 : index
    %device_0 = hal.devices.get %c0 : !hal.device
    %fence = hal.fence.create device(%device_0 : !hal.device) flags("None") : !hal.fence
    %1:5 = util.call @main$async(%arg0, %arg1, %arg2, %arg3, %0, %fence) : (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.fence, !hal.fence) -> (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view)
    %status = hal.fence.await until([%fence]) timeout_millis(%c-1_i32) flags("None") : i32
    util.return %1#0, %1#1, %1#2, %1#3, %1#4 : !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view
  }
}
```

### Normal Forward IR:

```mlir
#map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
module {
  util.func public @main$async(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view, %arg3: !hal.buffer_view, %arg4: !hal.fence, %arg5: !hal.fence) -> !hal.buffer_view attributes {inlining_policy = #util.inline.never, iree.abi.model = "coarse-fences", iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = hal.tensor.import wait(%arg4) => %arg0 : !hal.buffer_view -> tensor<2xf32>
    %1 = hal.tensor.import wait(%arg4) => %arg1 : !hal.buffer_view -> tensor<2x3x1x1xf32>
    %2 = hal.tensor.import wait(%arg4) => %arg2 : !hal.buffer_view -> tensor<2x3x16x16xf32>
    %3 = hal.tensor.import wait(%arg4) => %arg3 : !hal.buffer_view -> tensor<2x2x16x16xf32>
    %4 = tensor.empty() : tensor<2x2x16x16xf32>
    %broadcasted = linalg.broadcast ins(%0 : tensor<2xf32>) outs(%4 : tensor<2x2x16x16xf32>) dimensions = [0, 2, 3] 
    %5 = linalg.conv_2d_nchw_fchw {dilations = dense<1> : vector<2xi64>, strides = dense<1> : vector<2xi64>} ins(%2, %1 : tensor<2x3x16x16xf32>, tensor<2x3x1x1xf32>) outs(%broadcasted : tensor<2x2x16x16xf32>) -> tensor<2x2x16x16xf32>
    %6 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%5 : tensor<2x2x16x16xf32>) outs(%4 : tensor<2x2x16x16xf32>) {
    ^bb0(%in: f32, %out: f32):
      %10 = arith.cmpf ugt, %in, %cst : f32
      %11 = arith.select %10, %in, %cst : f32
      linalg.yield %11 : f32
    } -> tensor<2x2x16x16xf32>
    %7 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%6, %3 : tensor<2x2x16x16xf32>, tensor<2x2x16x16xf32>) outs(%4 : tensor<2x2x16x16xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %10 = arith.mulf %in, %in_0 : f32
      linalg.yield %10 : f32
    } -> tensor<2x2x16x16xf32>
    %8 = hal.tensor.barrier join(%7 : tensor<2x2x16x16xf32>) => %arg5 : !hal.fence
    %9 = hal.tensor.export %8 : tensor<2x2x16x16xf32> -> !hal.buffer_view
    util.return %9 : !hal.buffer_view
  }
  util.func public @main(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view, %arg3: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %0 = util.null : !hal.fence
    %c-1_i32 = arith.constant -1 : i32
    %c0 = arith.constant 0 : index
    %device_0 = hal.devices.get %c0 : !hal.device
    %fence = hal.fence.create device(%device_0 : !hal.device) flags("None") : !hal.fence
    %1 = util.call @main$async(%arg0, %arg1, %arg2, %arg3, %0, %fence) : (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.fence, !hal.fence) -> !hal.buffer_view
    %status = hal.fence.await until([%fence]) timeout_millis(%c-1_i32) flags("None") : i32
    util.return %1 : !hal.buffer_view
  }
}
```

You can reproduce by running:

```
TURBINE_LOG_LEVEL=DEBUG pytest tests/dynamo/backend_test.py -vv
```

Or by compiling both for `--iree-hal-target-backends=rocm --iree-hip-target=gfx942` and running on splat 1.0 inputs to compare the results of the first output tensor.

**Comment by zjgarvey** [comment]
For testing `iree-run-module`, I did the following (which gave the correct output of splat 0.1):

### Save IR above to `grads.mlir` and `no_grads.mlir`, respectively

### Compile each mlir file for gfx942

```
iree-compile --iree-hal-target-backends=rocm --iree-hip-target=gfx942 no_grads.mlir -o no_grads.vmfb
iree-compile --iree-hal-target-backends=rocm --iree-hip-target=gfx942 grads.mlir -o grads.vmfb
```

### Run with iree-run-module

```
iree-run-module --device=hip --module=no_grads.vmfb --input="2xf32=-2.5" --input="2x3x1x1xf32=1.0" --input="2x3x16x16xf32=1.0" --input="2x2x16x16xf32=0.2"
iree-run-module --device=hip --module=grads.vmfb --input="2xf32=-2.5" --input="2x3x1x1xf32=1.0" --input="2x3x16x16xf32=1.0" --input="2x2x16x16xf32=0.2"
```

## By comparison:

With `iree_turbine` installed in your python environment (and including commits from https://github.com/iree-org/iree-turbine/pull/953):

```python
import torch

@torch.compile(backend="iree_turbine")
def conv_relu_mul(x, y, z, w):
    conv = torch.conv2d(x, y, z)
    relu = torch.nn.functional.relu(conv)
    return relu*w

x = torch.ones([2, 3, 16, 16], dtype=torch.bfloat16, device="cuda:0", requires_grad=True)
y = torch.ones([2, 3, 1, 1], dtype=torch.bfloat16, device="cuda:0", requires_grad=True)
z = (-2.5)*torch.ones([2], dtype=torch.bfloat16, device="cuda:0", requires_grad=True)
w = 0.2*torch.ones([2,2,16,16], dtype=torch.bfloat16, device="cuda:0", requires_grad=True)

output = conv_relu_mul(x,y,z,w)
print(output)
```

Running this python script returns for me a splat tensor of `0.5`, which is not correct (it should be 0.1). E.g. compare the result of commenting out `@torch.compile...` or removing the `requires_grad=True` for each input.

**Comment by AWoloszyn** [comment]
So initial investigation into this:

GPU pointers returned by the invoke:
`0x70d098200000,0x70d099e01800,0x70d099e00000,0x70d099e02e00,0x70d098200000`
CPU pointers returned by the invoke
`0x370ad0c0,0x37700240,0x37621ec0,0x375f8f40,0x370ac0c0`

Importantly the FIRST and LAST pointers on the CPU-side are not the same, the FIRST and LAST pointers on the GPU-side ARE. So it looks like we are either returning the wrong buffer view, or (more likely) exporting the wrong range of the buffer view. Investigating now.
